reference-library/BeforeSwapDelta.md

```diff
@@ -0,0 +1,218 @@
+{
+    // ... (other code)
+
+    BeforeSwapDelta beforeSwapDelta;
+    {
+        int256 amountToSwap;
+        uint24 lpFeeOverride;
+        (amountToSwap, beforeSwapDelta, lpFeeOverride) = key.hooks.beforeSwap(key, params, hookData);
+
+        // ... (swap execution)
+    }
+
+    // ... (other code)
+}
+```
+
+The `beforeSwap` hook returns a `BeforeSwapDelta` value along with other parameters.
+
+- This `beforeSwapDelta` can then be passed to the `afterSwap` hook:
+
+```solidity
+BalanceDelta hookDelta;
+(swapDelta, hookDelta) = key.hooks.afterSwap(key, params, swapDelta, hookData, beforeSwapDelta);
+```
+
+The `BeforeSwapDelta` serves several important purposes in this context:
+
+1. **Customization of Swap Behavior:** It allows hooks to modify the swap parameters or even completely override the default swap behavior.
+2. **Information Passing:** It provides a way for the `beforeSwap` hook to pass information to the `afterSwap` hook, enabling more complex and stateful hook logic.
+3. **Balance Adjustment:** The delta values can be used to adjust the final balance changes resulting from the swap, giving hooks fine-grained control over the swap's outcome.
+4. **Gas Optimization:** By packing two `int128` values into a single `int256`, it reduces the number of stack variables and can lead to gas savings.
+
+This usage of `BeforeSwapDelta` is central to the flexibility and extensibility of Uniswap V4's hook system, allowing for custom swap logic while maintaining a standardized interface for all pools.
+
+## Perspective
+
+It's important to note that the `BeforeSwapDelta` is from the perspective of the hook itself, not the user. For example, if a user swaps 1 USDC for 1 USDT:
+
+- User gives 1 USDC: balance0OfUser decreases
+- Hook gets 1 USDC: balance0OfHook increases
+
+This perspective is key to correctly interpreting and manipulating the delta values within hook implementations.
+
+## Implementation Details
+
+The `BeforeSwapDelta` type and its associated functions use low-level assembly code for efficient bit manipulation and gas optimization:
+
+- The `toBeforeSwapDelta` function uses bitwise operations (`shl`, `or`, `and`, `sub`) to pack two `int128` values into a single `int256`.
+- The `getSpecifiedDelta` function uses the `sar` (shift arithmetic right) operation to extract the upper 128 bits.
+- The `getUnspecifiedDelta` function uses the `signextend` operation to extract and sign-extend the lower 128 bits.
+
+The `toBeforeSwapDelta` function combines the specified and unspecified deltas into a single `int256` value using bitwise operations. The `getSpecifiedDelta` and `getUnspecifiedDelta` functions extract the respective deltas using bit shifting and sign extension.
+
+By leveraging this compact representation and efficient arithmetic operations, Uniswap V4 can perform complex balance calculations and updates in a gas-optimized manner, reducing the overall cost of executing pool-related operations.
```

Comment:
can leave in, not sure if we need this level of detail for other docs. the reference library is mostly for information on how and why to use things, not how its built 😄